### ISSUE TITLE
Excessive heap allocations looking up GcmSpec constructor

### ISSUE THREAD

**Issue by bsmith-tridium-com** [bcgit/bc-java#1326]
YourKit tracks that 33% of allocation frequency and 16% of allocation size is spent looking up the same constructor in the BouncyCastle GcmSpecUtil class.

See allocation details:
![image](https://user-images.githubusercontent.com/31111326/216102857-36ba7819-d63d-499c-9d24-8a11b9e7f844.png)

See the code where the allocation takes place (NOTE is from me).

```
    public static AlgorithmParameterSpec extractGcmSpec(ASN1Primitive spec)
        throws InvalidParameterSpecException
    {
        try
        {
            GCMParameters gcmParams = GCMParameters.getInstance(spec);
            //NOTE: This should be cached!
            Constructor constructor = gcmSpecClass.getConstructor(new Class[]{Integer.TYPE, byte[].class});            return (AlgorithmParameterSpec)constructor.newInstance(new Object[] { Integers.valueOf(gcmParams.getIcvLen() * 8), gcmParams.getNonce() });
        }
        catch (NoSuchMethodException e)
        {
            throw new InvalidParameterSpecException("No constructor found!");   // should never happen
        }
        catch (Exception e)
        {
            throw new InvalidParameterSpecException("Construction failed: " + e.getMessage());   // should never happen
        }
    } 
```
Under normal circumstances this might be less expensive; however, in our SecurityManager environment, it is an expensive operation:
![image](https://user-images.githubusercontent.com/31111326/216103268-65c68de8-0f6c-455d-9a8b-84672008230d.png)
One can imagine that BCTLS provider is not routinely tested in a SecurityManager enabled environment.

See a contemporary solution to this problem against latest state of the BC git repository:
```
diff --git a/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/util/GcmSpecUtil.java b/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/util/GcmSpecUtil.java
index 41a0379eb..1dd812202 100644
--- a/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/util/GcmSpecUtil.java
+++ b/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/util/GcmSpecUtil.java
@@ -21,6 +21,7 @@     static final Method tLen;
     static final Method iv;
+    static final Constructor constructor;     static
     {
@@ -28,11 +29,13 @@
         {
             tLen = extractMethod("getTLen");
             iv = extractMethod("getIV");
+            constructor = extractConstructor(new Class[]{Integer.TYPE, byte[].class});
         }
         else
         {
             tLen = null;
             iv = null;
+            constructor = null;
         }
     }@@ -54,6 +57,25 @@ public Object run()
             return null;
         }
     }
+
+    private static Constructor extractConstructor(final Class [] parameterType)
+    {
+        try
+        {
+            return (Constructor)AccessController.doPrivileged(new PrivilegedExceptionAction()
+            {
+                public Object run()
+                    throws Exception
+                {
+                    return gcmSpecClass.getConstructor(parameterType);
+                }
+            });
+        }
+        catch (PrivilegedActionException e)
+        {
+            return null;
+        }
+    }     public static boolean gcmSpecExists()
     {
@@ -76,14 +98,8 @@ public static AlgorithmParameterSpec extractGcmSpec(ASN1Primitive spec)
         try
         {
             GCMParameters gcmParams = GCMParameters.getInstance(spec);
-            Constructor constructor = gcmSpecClass.getConstructor(new Class[]{Integer.TYPE, byte[].class});
-
             return (AlgorithmParameterSpec)constructor.newInstance(new Object[] { Integers.valueOf(gcmParams.getIcvLen() * 8), gcmParams.getNonce() });
         }
-        catch (NoSuchMethodException e)
-        {
-            throw new InvalidParameterSpecException("No constructor found!");   // should never happen
-        }
         catch (Exception e)
         {
             throw new InvalidParameterSpecException("Construction failed: " + e.getMessage());   // should never happen
(END)
         }
     }
+
+    private static Constructor extractConstructor(final Class [] parameterType)
+    {
+        try
+        {
+            return (Constructor)AccessController.doPrivileged(new PrivilegedExceptionAction()
+            {
+                public Object run()
+                    throws Exception
+                {
+                    return gcmSpecClass.getConstructor(parameterType);
+                }
+            });
+        }
+        catch (PrivilegedActionException e)
+        {
+            return null;
+        }
+    }     public static boolean gcmSpecExists()
     {
@@ -76,14 +98,8 @@ public static AlgorithmParameterSpec extractGcmSpec(ASN1Primitive spec)
         try
         {
             GCMParameters gcmParams = GCMParameters.getInstance(spec);
-            Constructor constructor = gcmSpecClass.getConstructor(new Class[]{Integer.TYPE, byte[].class});
-
             return (AlgorithmParameterSpec)constructor.newInstance(new Object[] { Integers.valueOf(gcmParams.getIcvLen() * 8), gcmParams.getNonce() });
         }
-        catch (NoSuchMethodException e)
-        {
-            throw new InvalidParameterSpecException("No constructor found!");   // should never happen
-        }
         catch (Exception e)
         {
             throw new InvalidParameterSpecException("Construction failed: " + e.getMessage());   // should never happen 
```
A casual GC cycle comparison demonstrates improvement:
![image](https://user-images.githubusercontent.com/31111326/216104367-e4fdd696-94bd-4581-ac13-cf2763cf23ed.png)

**Comment by peterdettman** [comment]
I have applied more or less the suggested fix. The issue also affected BCJSSE separately, so I've fixed it there similarly; you might notice a double improvement when using BCJSSE.

I expect a new beta with this fix will be made available within a few days.

**Comment by bsmith-tridium-com** [comment]
Thx Peter!

**Comment by peterdettman** [comment]
New beta is up at https://downloads.bouncycastle.org/betas/ .

**Comment by peterdettman** [comment]
@bsmith-tridium-com Were you able to test the beta and compare to your earlier results?

**Comment by bsmith-tridium-com** [comment]
@peterdettman, yes this seemed to address the issue.

**Comment by dghgit** [comment]
Fixed in 1.73
